### PR TITLE
8270085: Suspend during block transition may deadlock if lock held

### PR DESCRIPTION
**Pull request by pchilano**
Hi all,

The following patch fixes deadlocks issues that could occur when checking for suspension while holding VM locks. See the bug description for a concrete case. The solution is to avoid checking for suspend requests when using the TBIVM wrapper. The original patch was actually written by @robehn(now on vacations) and I only made small changes.

Run tiers1-6 in mach5 and currently running tier7. I also verified that the new added test SuspendBlocked.java deadlocks with the current bits and passes with this patch.

Thanks,
 Patricio

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8270085](https://bugs.openjdk.java.net/browse/JDK-8270085): Suspend during block transition may deadlock if lock held

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 603c62aed3490b95982584f1f111059a9a08c29c
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4828/head:pull/4828` \
`$ git checkout pull/4828`

Update a local copy of the PR: \
`$ git checkout pull/4828` \
`$ git pull https://git.openjdk.java.net/jdk pull/4828/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4828`

View PR using the GUI difftool: \
`$ git pr show -t 4828`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4828.diff">https://git.openjdk.java.net/jdk/pull/4828.diff</a>

</details>

### CODE REVIEW

**Review by dholmes-ora — APPROVED** [review]
Hi Patricio,

Thanks for taking over this fix from @robehn! And thanks Robbin for the initial fix.

So in simple terms we must never suspend whilst holding a VM internal lock (or any critical resource). So the ThreadBlockInVM destructor, which runs after we acquired a contended lock, must not suspend, but the ThreadBlockInVMPreProcess that we use for ObjectMonitors must suspend (and will release the lock).

I took a look at the other uses for ThreadBlockInVM. There are quite a few places where ThreadBlockInVM is used, some of which relate to threads that can't be suspended, but others where it looks very suspicious that we might have suspended before this change - so I think this is addressing a number of potential bugs.

I do wonder (perhaps separate RFE as it touches a number of files) whether ThreadBlockInVM should now be renamed ThreadBlockInVMNoSuspend to make it very clear the thread can't suspend? Though that might also mislead people as some of the logic is executed by threads that don't allow for suspension (not visible to Java code in general).

Overall I think this fix looks good, but I can't help wonder if the switch to using handshakes for suspend, is not still causing suspension checks in far more places than we used to check, and in potentially inappropriate (or at best unnecessary) places ... I'll take that thought up off line.

Thanks,
David

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 1]
```diff
@@ -289,7 +291,7 @@ class ThreadBlockInVM {
   ThreadBlockInVMPreprocess<InFlightMutexRelease> _tbivmpp;
  public:
   ThreadBlockInVM(JavaThread* thread, Mutex** in_flight_mutex_addr = NULL)
-    : _ifmr(in_flight_mutex_addr), _tbivmpp(thread, _ifmr) {}
+    : _ifmr(in_flight_mutex_addr), _tbivmpp(thread, _ifmr, false) {}
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/runtime/interfaceSupport.inline.hpp]: Done.

File: test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright (c) 2020, 2021, Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (pchilano) [reply test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java]: Fixed.

File: test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            suspend_thread.resume();
+            Thread.sleep(1);
+        }
+        suspend_thread.join();
+    }
+
+    public static void run_loop() {
+        WhiteBox wb = WhiteBox.getWhiteBox();
+        for (int i = 0; i < 100; i++) {
+            wb.lockAndBlock(false /* suspender */);
```
[the remark this answers is not in the document]
Comment (pchilano) [reply test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java]: Fixed.

**Comment by pchilano** [comment]
Hi David,

> Hi Patricio,
> 
> Thanks for taking over this fix from @robehn! And thanks Robbin for the initial fix.
> 
> So in simple terms we must never suspend whilst holding a VM internal lock (or any critical resource). So the ThreadBlockInVM destructor, which runs after we acquired a contended lock, must not suspend, but the ThreadBlockInVMPreProcess that we use for ObjectMonitors must suspend (and will release the lock).
> 
> I took a look at the other uses for ThreadBlockInVM. There are quite a few places where ThreadBlockInVM is used, some of which relate to threads that can't be suspended, but others where it looks very suspicious that we might have suspended before this change - so I think this is addressing a number of potential bugs.
> 
> I do wonder (perhaps separate RFE as it touches a number of files) whether ThreadBlockInVM should now be renamed ThreadBlockInVMNoSuspend to make it very clear the thread can't suspend? Though that might also mislead people as some of the logic is executed by threads that don't allow for suspension (not visible to Java code in general).
> 
In general I think this details of what things get processed when transitioning from one state to another should not be exposed to the users unless is necessary. In this case I think most users of TBIVM don't really care about processing suspend requests or not other than TBIVM should do whatever is right, which here is just never process them. For the special cases of java monitors and jvmti raw monitors where we specifically want to honor suspend requests we have the special ThreadBlockInVMPreprocess. I actually looked into having a single TBIVM with an additional parameter to allow for suspend in those special cases. It wasn't straightforward because of the template but I'll try to revisit that.

Thanks for reviewing this David!

Patricio

**Comment by dcubed-ojdk** [comment]
This fix appears to be baselined on https://github.com/openjdk/jdk
instead of JDK17. The bug is targeted at JDK17 so you probably need
to reparent the PR (if possible) or recreate it (ouch).

I'm going to go ahead and start my review anyway.

**Comment by pchilano** [comment]
> This fix appears to be baselined on https://github.com/openjdk/jdk
> instead of JDK17. The bug is targeted at JDK17 so you probably need
> to reparent the PR (if possible) or recreate it (ouch).
> 
> I'm going to go ahead and start my review anyway.
Ok, I thought I was supposed to integrate it in 18 first and then backport it (?).

**Comment by dcubed-ojdk** [comment]
It's a P2 so you should request permission to integrate it in JDK17 (follow
the RDP2 rules) and integrate there first. It will be automatically forward
ported when @JesperIRL syncs from JDK17 -> JDK18.

File: src/hotspot/share/prims/whitebox.cpp [hunk 4]
```diff
@@ -2122,6 +2122,31 @@ WB_ENTRY(void, WB_AsyncHandshakeWalkStack(JNIEnv* env, jobject wb, jobject threa
...
+    // Otherwise if either suspender or suspendee blocks for a safepoint
+    // in ~ThreadBlockInVM the other one could loop forever trying to acquire
+    // the lock without allowing the safepoint to progress.
+    ThreadBlockInVM tbivm(self);
+
+    // We will deadlock here if we are 'suspender' and 'suspendee'
+    // suspended in ~ThreadBlockInVM. This verifies we only suspend
+    // at the right place.
+    while (Atomic::cmpxchg(&_emulated_lock, 0, 1) != 0) {}
+    assert(_emulated_lock == 1, "Not locked");
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/prims/whitebox.cpp]: Fixed.

File: src/hotspot/share/prims/whitebox.cpp [hunk 5]
```diff
@@ -2122,6 +2122,31 @@ WB_ENTRY(void, WB_AsyncHandshakeWalkStack(JNIEnv* env, jobject wb, jobject threa
...
+
+    // Sleep much longer in suspendee to force situation where
+    // 'suspender' is waiting above to acquire lock.
+    os::naked_short_sleep(suspender ? 1 : 10);
+  }
+  Atomic::store(&_emulated_lock, 0);
+WB_END
+
 //Some convenience methods to deal with objects from java
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/prims/whitebox.cpp]: Fixed.

File: src/hotspot/share/prims/whitebox.cpp [hunk 6]
```diff
@@ -2613,6 +2638,9 @@ static JNINativeMethod methods[] = {
   {CC"handshakeReadMonitors", CC"(Ljava/lang/Thread;)Z", (void*)&WB_HandshakeReadMonitors },
   {CC"handshakeWalkStack", CC"(Ljava/lang/Thread;Z)I", (void*)&WB_HandshakeWalkStack },
   {CC"asyncHandshakeWalkStack", CC"(Ljava/lang/Thread;)V", (void*)&WB_AsyncHandshakeWalkStack },
+
+  {CC"lockAndBlock", CC"(Z)V", (void*)&WB_LockAndBlock},
+
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/prims/whitebox.cpp]: Removed.

File: src/hotspot/share/runtime/handshake.cpp [hunk 7]
```diff
@@ -441,6 +451,11 @@ bool HandshakeState::have_non_self_executable_operation() {
   return _queue.contains(non_self_queue_filter);
 }
 
+bool HandshakeState::has_none_suspend_operation() {
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/runtime/handshake.cpp]: Fixed.

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp [hunk 8]
```diff
@@ -289,7 +291,7 @@ class ThreadBlockInVM {
   ThreadBlockInVMPreprocess<InFlightMutexRelease> _tbivmpp;
  public:
   ThreadBlockInVM(JavaThread* thread, Mutex** in_flight_mutex_addr = NULL)
-    : _ifmr(in_flight_mutex_addr), _tbivmpp(thread, _ifmr) {}
+    : _ifmr(in_flight_mutex_addr), _tbivmpp(thread, _ifmr, false /* no suspend */) {}
```
[the remark this answers is not in the document]
Comment (pchilano) [reply src/hotspot/share/runtime/interfaceSupport.inline.hpp]: Fixed.

File: src/hotspot/share/runtime/handshake.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (coleenp) [reply src/hotspot/share/runtime/handshake.cpp]: Me too.

File: src/hotspot/share/runtime/handshake.cpp [hunk 9]
```diff
@@ -667,8 +680,15 @@ class SuspendThreadHandshake : public HandshakeClosure {
...
 bool HandshakeState::suspend() {
+  JavaThread* self = JavaThread::current();
   SuspendThreadHandshake st;
   Handshake::execute(&st, _handshakee);
+  if (_handshakee == self) {
+    // If target is the current thread we need to call this to do the
+    // actual suspend since Handshake::execute() above only installed
+    // the asynchronous handshake.
+    SafepointMechanism::process_if_requested(self);
+  }
```
Comment (coleenp) [inline src/hotspot/share/runtime/handshake.cpp:691]: Is this an optimization?  Or can the thread escape the suspend request?

File: test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test SuspendBlocked
+ * @bug 8270085
+ * @library /testlibrary /test/lib
```
Comment (coleenp) [inline test/hotspot/jtreg/runtime/handshake/SuspendBlocked.java:28]: I think this is referring to /testlibrary that I removed.  Can you see if it's still needed?

**Review by coleenp — APPROVED** [review]
Just some comments and questions.  Looks good!

**Comment by pchilano** [comment]
jdk17 PR: https://github.com/openjdk/jdk17/pull/257

**Comment by pchilano** [comment]
Closing and using jdk17 PR: https://github.com/openjdk/jdk17/pull/257

File: src/hotspot/share/runtime/interfaceSupport.inline.hpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (dholmes-ora) [reply src/hotspot/share/runtime/interfaceSupport.inline.hpp]: No I said there were two styles for doing this kind of commenting. I like this style and suggested the comment.

**Pull request closed without merging**